### PR TITLE
Document api module for 1.0

### PR DESCRIPTION
**Pull request by arturbosch**
- document remaining files
- move helper functions to internal or other modules

### CODE REVIEW

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/AnnotationExcluder.kt [hunk 1]
```diff
@@ -3,18 +3,35 @@ package io.gitlab.arturbosch.detekt.api
 import org.jetbrains.kotlin.psi.KtAnnotationEntry
 import org.jetbrains.kotlin.psi.KtFile
 
+/**
+ * Primary use case for an AnnotationExcluder is to test KtElement's
+ * if they have an annotation which is excluded e.g. okay for a given rule
+ * not to report it.
```
Comment (schalkms) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/AnnotationExcluder.kt]: Could we please rephrase this?
`if they have an annotation which excludes the element from being reported` or sth like that

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/AnnotationExcluder.kt [hunk 2]
```diff
@@ -3,18 +3,35 @@ package io.gitlab.arturbosch.detekt.api
...
 			?.filterNot { it.isAllUnder }
 			?.mapNotNull { it.importedFqName?.asString() }
-			?.map { Pair(it.split(".").last(), it) }
+			?.map { Pair(it.substringAfterLast('.'), it) }
 			?.toMap()
 
+	/**
+	 * Is true if any given annotation name is declared in the SplitPattern
+	 * which basically describes entries to exclude.
+	 */
```
Comment (schalkms) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/AnnotationExcluder.kt]: add `@param` `@return` doc here?

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/BaseRule.kt [hunk 3]
```diff
@@ -26,19 +42,27 @@ abstract class BaseRule(protected val context: Context = DefaultContext()) : Det
 		root.accept(this)
 	}
 
+	/**
+	 * Basic mechanism to decide if a rule should run or not.
+	 *
+	 * By default any rule which is declared 'active' in the [Config]
+	 * or not suppressed by a [Suppress] annotation on file level should run.
+	 */
```
Comment (schalkms) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/BaseRule.kt]: add `@param` `@return` doc here?

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/BaseRule.kt [hunk 4]
```diff
@@ -26,19 +42,27 @@ abstract class BaseRule(protected val context: Context = DefaultContext()) : Det
...
 	abstract fun visitCondition(root: KtFile): Boolean
 
 	/**
 	 * Could be overridden by subclasses to specify a behaviour which should be done before
 	 * visiting kotlin elements.
 	 */
-	protected open fun postVisit(root: KtFile) {
+	protected open fun preVisit(root: KtFile) {
```
Comment (schalkms) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/BaseRule.kt]: add `@param` doc here?

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/BaseRule.kt [hunk 5]
```diff
@@ -26,19 +42,27 @@ abstract class BaseRule(protected val context: Context = DefaultContext()) : Det
...
+		// nothing to do by default
 	}
 
 	/**
 	 * Could be overridden by subclasses to specify a behaviour which should be done after
 	 * visiting kotlin elements.
 	 */
-	protected open fun preVisit(root: KtFile) {
+	protected open fun postVisit(root: KtFile) {
```
Comment (schalkms) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/BaseRule.kt]: add `@param` doc here?

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Entity.kt [hunk 6]
```diff
@@ -17,6 +20,9 @@ data class Entity(val name: String,
 	override fun compact() = "[$name] at ${location.compact()}"
 
 	companion object {
+		/**
+		 * Factory function which retrieves all needed information from the [PsiElement] itself.
+		 */
 		fun from(element: PsiElement, offset: Int = 0): Entity {
```
Comment (schalkms) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Entity.kt]: add `@param` `@return` doc here?

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/OutputReport.kt [hunk 7]
```diff
@@ -4,16 +4,28 @@ import java.nio.file.Files
...
+	/**
+	 * Name of the report. Is used to exclude this report in the yaml config.
+	 */
 	open val name
 		get() = this::class.simpleName
 
+	/**
+	 * Renders result and writes it to the given [filePath].
+	 */
 	fun write(filePath: Path, detektion: Detektion) {
```
Comment (schalkms) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/OutputReport.kt]: add `@param` doc here?

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/OutputReport.kt [hunk 8]
```diff
@@ -25,5 +37,8 @@ abstract class OutputReport : Extension {
 		}
 	}
 
+	/**
+	 * Defines the translation process of detekt's result into a string.
+	 */
 	abstract fun render(detektion: Detektion): String?
```
Comment (schalkms) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/OutputReport.kt]: add `@param` `@return` doc here?

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Rule.kt [hunk 9]
```diff
@@ -46,10 +46,16 @@ abstract class Rule(override val ruleSetConfig: Config = Config.empty,
 
 	override fun visitCondition(root: KtFile) = active && !root.isSuppressedBy(ruleId, aliases)
 
+	/**
+	 * Simplified version of [Context.report] with aliases retrieval from the config.
+	 */
 	fun report(finding: Finding) {
```
Comment (schalkms) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Rule.kt]: add `@param` doc here?

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/SingleAssign.kt [hunk 10]
```diff
@@ -3,6 +3,9 @@ package io.gitlab.arturbosch.detekt.api
 import kotlin.reflect.KProperty
 
 /**
+ * Allows to assign a property just one time.
```
Comment (schalkms) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/SingleAssign.kt]: just once

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/internal/Signatures.kt [hunk 11]
```diff
@@ -88,3 +88,19 @@ private fun buildFunctionSignature(element: KtNamedFunction): String {
 			{ element.nameAsSafeName.identifier },
 			{ element.text.substring(methodStart, methodEnd) })
 }
+
+/*
+ * When analyzing sub path 'testData' of the kotlin project, CompositeElement.getText() throws
+ * a RuntimeException stating 'Underestimated text length' - #65.
+ */
+@Suppress("TooGenericExceptionCaught")
+internal fun getTextSafe(defaultValue: () -> String, block: () -> String) = try {
```
Comment (schalkms) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/internal/Signatures.kt]: add `@param` `@return` doc here?

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/IsPartOfUtils.kt [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+import kotlin.reflect.KClass
+
+/**
+ * @author Artur Bosch
+ */
+
+/**
+ * Tests if this element is part of given PsiElement.
+ */
+fun PsiElement.isPartOf(clazz: KClass<out PsiElement>) = getNonStrictParentOfType(clazz.java) != null
```
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/IsPartOfUtils.kt]: add `@param` `@return` doc here?

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/IsPartOfUtils.kt [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+
+/**
+ * Tests of this element is part of a kotlin string.
+ */
+fun PsiElement.isPartOfString() = isPartOf(KtStringTemplateEntry::class)
```
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/IsPartOfUtils.kt]: add `@return` doc here?

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/AnnotationExcluder.kt (hunk 2 above)
Comment (arturbosch) [reply detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/AnnotationExcluder.kt]: I think adding `@param` and `@return` add little to no value.
Looking at the classes provides better insights. 
However we should always specify the return type `: Boolean` here.

Comment (schalkms) [reply detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/AnnotationExcluder.kt]: It's common practice to which detekt might adhere to?

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/IsPartOfUtils.kt [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+ * @author Artur Bosch
+ */
+
+/**
+ * Tests if this element is part of given PsiElement.
+ */
+fun PsiElement.isPartOf(clazz: KClass<out PsiElement>) = getNonStrictParentOfType(clazz.java) != null
+
+/**
+ * Tests of this element is part of a kotlin string.
```
Comment (3flex) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/IsPartOfUtils.kt]:
```suggestion
 * Tests if this element is part of a Kotlin string.
```

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/AnnotationExcluder.kt (hunk 2 above)
Comment (schalkms) [reply detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/AnnotationExcluder.kt]: Ok good to know.

**Pull request merged**